modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/streaming/PagingProviderProducer.java

```diff
@@ -77,16 +88,35 @@ public int getSize() {
    * @return
    */
   private <R> R performWithConnection(Function<Object, R> function) {
+    Optional<MutableConfigurationStats> stats = getMutableConfigurationStats(executionContext);
+    RetryPolicyTemplate retryPolicy =
+        (RetryPolicyTemplate) executionContext.getRetryPolicyTemplate().orElseGet(NoRetryPolicyTemplate::new);
+    CompletableFuture<R> future = retryPolicy.applyPolicy(() -> completedFuture(withConnection(function)),
+                                                          e -> shouldRetry(e, executionContext),
+                                                          e -> {
+                                                          },
+                                                          e -> stats.ifPresent(s -> s.discountInflightOperation()),
```

Comment:
This is repeated code from the mediator... everything between lines 92 and 100 should be part of the ReconnectionUtils class you created as a static method that receives the CompletableFuture and is used both here and in the mediator